lib/Providers/OpenIDConnectProvider.php

```diff
@@ -0,0 +1,186 @@
+        if (!empty($data[$this->responseError])) {
+            $error = $data[$this->responseError];
+            if (!is_string($error)) {
+                $error = var_export($error, true);
+            }
+        }
+        if ($error || $response->getStatusCode() >= 400) {
+            throw new IdentityProviderException($error, 0, $data);
+        }
+    }
+
+    protected function createResourceOwner(array $response, AccessToken $token)
+    {
+        return new GenericResourceOwner($response, 'id');
+    }
+
+    /**
+     * Do any required verification of the id token and return an array of decoded claims
+     *
+     * @param string $id_token Raw id token as string
+     * @return array associative array of claims decoded from the id token
+     */
+    public function verifyIdToken($id_token)
+    {
+        try {
+            $keys = $this->getSigningKeys();
+            $claims = JWT\JWT::decode($id_token, $keys, ['RS256']);
+            if ($claims->aud !== $this->clientId) {
+                throw new IdentityProviderException("ID token has incorrect audience", 0, $claims->aud);
+            }
+            if ($claims->iss !== $this->issuer) {
+                throw new IdentityProviderException("ID token has incorrect issuer", 0, $claims->iss);
+            }
+        } catch (\UnexpectedValueException $e) {
+            throw new IdentityProviderException("ID token validation failed", 0, $e->getMessage());
+        }
+    }
+
+    protected function prepareAccessTokenResponse(array $result)
+    {
+        $result = parent::prepareAccessTokenResponse($result);
+        $this->verifyIdToken($result['id_token']);
+        return $result;
+    }
+
+    protected function getOpenIDConfiguration() {
+        if (isset($this->openIdConfiguration)) {
+            return $this->openIdConfiguration;
+        }
+
+        $config = $this->getParsedResponse($this->getRequest('GET', $this->issuer . self::CONFIGURATION_PATH));
+        $this->openIdConfiguration = $config;
+        return $config;
+    }
+
+    public static function base64url_decode($input) {
+        return base64_decode(strtr($input, '-_', '+/'));
+    }
+
+    protected function getSigningKeys() {
+        $url = $this->getOpenIDConfiguration()['jwks_uri'];
+        $jwks = $this->getParsedResponse($this->getRequest('GET', $url));
+        $keys = [];
+        foreach ($jwks['keys'] as $key) {
+            $kid = $key['kid'];
+            if (array_key_exists('x5c', $key)) {
+                $x5c = $key['x5c'];
+                $keys[$kid] = "-----BEGIN CERTIFICATE-----\n" . $x5c[0] . "\n-----END CERTIFICATE-----";
+            } else if ($key['kty'] === 'RSA') {
+                $e = self::base64url_decode($key['e']);
+                $n = self::base64url_decode($key['n']);
+                $keys[$kid] = \RobRichards\XMLSecLibs\XMLSecurityKey::convertRSA($n, $e);
+            } else {
+                error_log("Failed to load key data for key id: " . $kid);
```

Comment:
Logging should be done through `SimpleSAML\Logger` interface.

Reply:
My bad, copied this from the cli script a bit too quickly.